torch/utils/cpp_extension.py

```diff
@@ -2268,20 +2454,48 @@ def _write_ninja_file_to_build_library(path,
     else:
         cuda_flags = None
 
+    if with_sycl:
+        sycl_common_cflags = ['-fsycl', '-fsycl-targets=spir64_gen,spir64']
+        sycl_cflags = cflags + sycl_common_cflags
+        sycl_cflags += extra_sycl_cflags
+        # "preview breaking changes" flags allows to support both C++ ABIs
+        sycl_cflags += ['-fpreview-breaking-changes', '-D__INTEL_PREVIEW_BREAKING_CHANGES']
+        if not any(flag.startswith('-sycl-std=') for flag in sycl_cflags):
+            sycl_cflags.append('-sycl-std=2020')
+        host_cxx = get_cxx_compiler()
+        host_cflags = cflags
+        # escaping quoted arguments to pass them thru icpx
+        host_cflags = [item.replace('\\"', '\\\\"') for item in host_cflags]
+        host_cflags = ' '.join(host_cflags)
+        # note that -fsycl-host-compiler-options will be quoted as needed
+        # with shlex.quote below
+        sycl_cflags += [f'-fsycl-host-compiler={host_cxx}', shlex.quote(f'-fsycl-host-compiler-options={host_cflags}')]
+        sycl_arch_list = os.environ.get('TORCH_XPU_ARCH_LIST', 'pvc,xe-lpg')
+        sycl_dlink_post_cflags = sycl_common_cflags + ['-fsycl-link', f'-Xs "-device {sycl_arch_list}"']
+    else:
+        sycl_cflags = None
+        sycl_dlink_post_cflags = None
```

Comment:
Could we reuse the code block  - https://github.com/pytorch/pytorch/pull/132945/files#diff-258d5272916c8cf5ac89aad77f7c114b585848826f6bbaeee91fe62a4391ee5eR727-R751?

Reply:
Well, that's challenging. As much as I wished to reuse, unfortunately these code blocks are different. They correspond to 2 different Extension APIs: one for `torch.utils.cpp_extension.load()` and anther for `class SyclExtension`. And these paths are implemented differently (same for cuda and cpp).

Additional difficulty is `-fsycl-host-compiler-options` option which is qouted space containing string which should be properly escaped. And code to correctly escape it is different for 2 Extension API scenarios noted above.

I have tried to improve this code block - see few new commits. I can't 100% reuse the code due to above reasons, but at least I moved definition of common flags to the global scope which allowed to simplify code a bit and somewhat reduce duplication.